RPC: Add parameter to addmultisigaddress / createmultisig to sort public keys

afk11 opened the pull request:
I figured it may be useful for these RPC methods to allow sorting public keys (BIP67) The PR adds a new boolean to createmultisig / addmultisigaddress at the end of their parameter list. By default, this is set to false to avoid a BC break. 

I added a RPC test file `sort_multisig.py` for testing createmultisig. Tests for addmultisigaddress went in `wallet-accounts.py`.

Note: Code to check whether sorting is desired had to be replicated in both RPC methods (not in _createmultisig_redeemScript) because addmultisigaddress already takes a parameter at position 3.

dcousens commented:
concept ACK

laanwj commented:
Concept ACK, although I really don't like multiple-optional-boolean APIs. Wish we switched to named arguments any day.

One nit: the RPC help should mention BIP67 by name.

MarcoFalke commented:
Needs rebase

afk11 committed:
RPC: addmultisigaddress / createmultisig: parameterize _createmultisig_redeemScript to allow sorting of public keys (BIP67)

addmultisig/createmultisig RPC documentation: Remove stray quotes from fSort parameter

afk11 commented:
@MarcoFalke thanks, done. @laanwj I should have mentioned, nits addressed. 

One travis run failed due to the compactblocks RPC test.

ryanofsky commented:
I can't see anything on travis right now  (503 errors), but the compactblocks error is probably just the spurious https://github.com/bitcoin/bitcoin/issues/8842 / https://github.com/bitcoin/bitcoin/pull/9058 failures.

MarcoFalke commented on qa/rpc-tests/sort_multisig.py:
```diff
@@ -0,0 +1,41 @@
+class SortMultisigTest(BitcoinTestFramework):
+    def __init__(self):
+        super().__init__()
+        self.num_nodes = 4
```
Nit: A single node should be enough?

MarcoFalke commented:
Concept ACK 7d7a64726991ff087cb8125e0c7277173a688dc7

luke-jr commented on src/rpc/misc.cpp:
```diff
@@ -293,6 +294,7 @@ UniValue createmultisig(const JSONRPCRequest& request)
             "       \"key\"    (string) bitcoin address or hex-encoded public key\n"
             "       ,...\n"
             "     ]\n"
+            "3. \"fSort\"      (bool, optional) Whether to sort public keys according to BIP67. Default setting is false.\n"
```
Is it a string or a boolean?

afk11 replied:
It should be a boolean. Just observed they aren't usually quoted in RPC output, fixing now.

afk11 commented:
I probably shouldn't have squashed @MarcoFalke, I'm sorry for rebasing out the commit you reviewed. The only thing to change this time was the removal of "'s from the RPC help message.

luke-jr commented on src/script/standard.cpp:
```diff
@@ -290,13 +290,24 @@ CScript GetScriptForRawPubKey(const CPubKey& pubKey)
-    CScript script;
+    int nEncoded = 0;
+    std::vector<std::vector<unsigned char>> vEncoded;
+    vEncoded.resize(keys.size());
```
Wouldn't it be better to `reserve` and then `emplace_back`?

luke-jr commented on src/script/standard.cpp:
```diff
@@ -290,13 +290,24 @@ CScript GetScriptForRawPubKey(const CPubKey& pubKey)
+    }
+
+    if (fSorted) {
+        std::sort(vEncoded.begin(), vEncoded.end());
```
I *think* this should do what BIP 67 requires, but someone more familiar with C++ and its locale support (or lack thereof) should probably confirm.

luke-jr commented on src/script/standard.h:
```diff
@@ -78,7 +78,7 @@ bool ExtractDestinations(const CScript& scriptPubKey, txnouttype& typeRet, std::
 CScript GetScriptForDestination(const CTxDestination& dest);
 CScript GetScriptForRawPubKey(const CPubKey& pubkey);
-CScript GetScriptForMultisig(int nRequired, const std::vector<CPubKey>& keys);
+CScript GetScriptForMultisig(int nRequired, const std::vector<CPubKey>& keys, bool fSorted);
```
Maybe default fSorted to false here rather than modify all the tests?

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -973,28 +973,30 @@ UniValue sendmany(const JSONRPCRequest& request)
             "       ...,\n"
             "     ]\n"
             "3. \"account\"      (string, optional) DEPRECATED. An account to assign the addresses to.\n"
+            "4. fSort          (bool, optional) Whether to sort public keys according to BIP67. Default setting is false.\n"
```
As before, rather turn param 3 into an options Object.

luke-jr reviewed (changes requested):
Code looks reasonably correct, just a few nits. Did not verify tests.

luke-jr commented:
Should update `doc/bips.md` also.

luke-jr committed:
script: GetScriptForRawPubKey: More efficient key-encoding loop

luke-jr committed:
doc/bips: Add BIP 67

luke-jr commented:
I rebased and addressed all the nits; pushed this to luke-jr/sort-multisigs

@afk11 Are you still maintaining this? Can you pull my changes?

```
git checkout sort-multisigs
git fetch git://github.com/luke-jr/bitcoin sort-multisigs
git reset --hard FETCH_HEAD
git push ...
```

afk11 commented:
Sorry, yep I can pull these! 

I wanted to wait until named parameters was merged before hand, so I could avoid adding a positional parameter before the accounts parameters were changed 

I'll look at this in the next day or so (away from internet atm) wanna finish this up

afk11 commented:
Merged commits and rebased. Apologies for the delay!

afk11 commented:
The Apple build failed because the job time exceeded the maximum :/

afk11 commented:
Rebased

afk11 commented:
Rebased

luke-jr commented:
The description for d70583a2f9ae8bb4bda2e6224e699ac2cbb583ee is no longer correct.

afk11 commented:
@luke-jr Oops, yeah the rebase mightn't have been the easiest. So re-add the contents of setup_network?

jnewbery commented:
Perhaps I'm missing something, but I don't see the need for this. The `addmultisigaddress` RPC creates the multisig script with the keys in the order provided. Why not just have the client provide keys in sorted order if you want the script to be BIP-67 compliant?

It doesn't look like this PR is enforcing that the provided keys are compressed, so even with this PR, there are still expectations placed on the client.

afk11 commented:
I think if developers are already committing to using the RPC to make a multisig script, making it easier to produce the same representation is more important than not.

You are correct the PR as it stands doesn't validate it.. fixing this now.

jnewbery commented:
I'm still a weak concept NACK for this. I don't agree that we should add complexity to the server when the same outcome can be achieved by simply running a `sort()` function on the client before calling the RPC. Sometimes there's good reason to add that complexity to the server - see for example #9991 which adds a filter to save significantly on bandwidth and server resources. In this case I don't see the benefit.

Sorry if that sounds negative - I think there needs to be some bar for adding new RPCs and arguments to avoid bloat.

However, if I'm wrong and there's widespread consensus that this is useful functionality and should be merged, can I at least ask that you use named arguments instead of an Options object? There's really no need for Options objects in RPC calls since named args were added in #8811.

afk11 commented:
Both RPC methods take an options object for this (sorry, the PR description wasn't updated with this)
https://github.com/bitcoin/bitcoin/pull/8751/files#diff-ad6efdc354b57bd1fa29fc3abb6e2872R353
https://github.com/bitcoin/bitcoin/pull/8751/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR1050 

I appreciate where you are coming from and agree that most people can probably sort themselves, but they could also build a multisig script out of the keys and m/n. It's been a while since I've even used the RPC, but remember well the time when I didn't have a bitcoin library to do it all. 

I think it's worth including since once they continue using the flag, requests which mistakenly use the wrong order will reproduce the same redeem script (instead of always having a stateful order of public keys), and likewise with libraries that support it.

ipoptika commented:
https://minecloud.io/?in=ipoptika


>Понедельник, 19 июня 2017, 11:53 +03:00 от Thomas Kerin <notifications@github.com>:
>
>Both RPC methods take an options object for this (sorry, the PR description wasn't updated with this)
>https://github.com/bitcoin/bitcoin/pull/8751/files#diff-ad6efdc354b57bd1fa29fc3abb6e2872R353
>https://github.com/bitcoin/bitcoin/pull/8751/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR1050
>I appreciate where you are coming from and agree that most people can probably sort themselves, but they could also build a multisig script out of the keys and m/n. It's been a while since I've even used the RPC, but remember well the time when I didn't have a bitcoin library to do it all.
>I think it's worth including since once they continue using the flag, requests which mistakenly use the wrong order will reproduce the same redeem script (instead of always having a stateful order of public keys), and likewise with libraries that support it.
>—
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly,  view it on GitHub , or  mute the thread .

luke-jr commented on src/script/standard.cpp:
```diff
@@ -288,13 +288,23 @@ CScript GetScriptForRawPubKey(const CPubKey& pubKey)
-    CScript script;
+    std::vector<std::vector<unsigned char>> vEncoded;
+    vEncoded.reserve(keys.size());
+    BOOST_FOREACH(const CPubKey& key, keys) {
```
We're not using `BOOST_FOREACH` anymore I think. (Also below)

luke-jr commented:
Rebased and squashed a bit.

```
git checkout sort-multisigs && git fetch git://github.com/luke-jr/bitcoin sort-multisigs && git reset --hard FETCH_HEAD && git push ...
```

TheBlueMatt commented:
Concept ACK. Care to rebase?

Commit:
Ensure whilst sorting only compressed public keys are used

Commit:
Add more tests to sort_multisigs.py / wallet-accounts.py

sort_multisig test: check uncompressed keys are disallowed
sort_multisig: add test demonstrating sorting
wallet-accounts: test addmultisigaddress fails if sort=true and (wallet) address is uncompressed

afk11 commented:
Rebased, sorry for the delay. Updated to check that keys are compressed before allowing sorting, and added more tests for this. 

Updated the docs/bips.md document to mention 0.15.1 instead of 0.15.0 (let me know whatever's best for this)

TheBlueMatt commented:
Hmm, hate to reopen it, but now that we do actually have named arguments, could you rever to just adding a new boolean argument? options objects are just redundant now, and having options alias account in addmultisigaddress is just gross. Everything else seems fine at first glance.

@jnewbery I'd generally agree with you, but, at least in principal, I think BIP67 is worth it.

afk11 commented:
@TheBlueMatt that's fine, revised the PR now. 

I missed the boat again for v0.15.1, suggestions for a release to mention in bips.md?

luke-jr commented:
> Hmm, hate to reopen it, but now that we do actually have named arguments, could you rever to just adding a new boolean argument? options objects are just redundant now, and having options alias account in addmultisigaddress is just gross. Everything else seems fine at first glance.

Strongly disagree. Named arguments is not a reason to have a terrible positional arguments API. Uncommon options should go through an options argument when positional arguments are used.

The account alias is merely for backward compatibility.

laanwj commented:
Closing and putting "up for grabs" label
